[Mainnet/Testnet] Add Decimals Property

mrtpain opened the pull request:
Adds a decimals property to the standard token contract, used for display and calculation purposes. 

```text
59089b9e88036b80c74b92d1232503d167c67ef5ef96934f741467442afac156
```

**Bytecode**

```text
4D5A90000300000004000000FFFF0000B800000000000000400000000000000000000000000000000000000000000000000000000000000000000000800000000E1FBA0E00B409CD21B8014CCD21546869732070726F6772616D2063616E6E6F742062652072756E20696E20444F53206D6F64652E0D0D0A2400000000000000504500004C0102003B5C9AD40000000000000000E00022200B013000000E000000020000000000001E2D0000002000000040000000000010002000000002000004000000000000000400000000000000006000000002000000000000030040850000100000100000000010000010000000000000100000000000000000000000CC2C00004F000000000000000000000000000000000000000000000000000000004000000C000000B02C00001C0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000080000000000000000000000082000004800000000000000000000002E74657874000000240D000000200000000E000000020000000000000000000000000000200000602E72656C6F6300000C000000004000000002000000100000000000000000000000000000400000420000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002D00000000000048000000020005007C230000340900000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000E20203280500000A0204280900000602052805000006020E042803000006020E0528070000060202280600000A6F0700000A04280B0000062A4602280800000A72010000706F0900000A2A4A02280800000A7201000070036F0A00000A2A4602280800000A720F0000706F0900000A2A4A02280800000A720F000070036F0A00000A2A4602280800000A72190000706F0B00000A2A4A02280800000A7219000070036F0C00000A2A4602280800000A722B0000706F0D00000A2A4A02280800000A722B000070036F0E00000A2A7202280800000A7243000070038C08000001280F00000A6F0D00000A2A7602280800000A7243000070038C08000001280F00000A046F0E00000A2A13300400A600000001000011042D34021201FE1503000002120102280600000A6F0700000A7D010000041201037D020000041201166A7D0300000407280100002B172A0202280600000A6F0700000A280A0000060A06043402162A0202280600000A6F0700000A0604DB280B00000602030203280A00000604D7280B000006021201FE1503000002120102280600000A6F0700000A7D010000041201037D020000041201047D0300000407280100002B172A000013300500AA00000002000011052D2A021202FE15030000021202037D010000041202047D020000041202166A7D0300000408280100002B172A020302280600000A6F0700000A28100000060A0203280A0000060B0605370407053402162A020302280600000A6F0700000A0605DB280F00000602030705DB280B00000602040204280A00000605D7280B000006021202FE15030000021202037D010000041202047D020000041202057D0300000408280100002B172A00001330040065000000030000110202280600000A6F0700000A032810000006042E02162A0202280600000A6F0700000A0305280F000006021200FE1504000002120002280600000A6F0700000A7D040000041200037D050000041200057D070000041200047D0600000406280200002B172A8E02280800000A725B000070038C08000001048C08000001281100000A056F0E00000A2A8A02280800000A725B000070038C08000001048C08000001281100000A6F0D00000A2A42534A4201000100000000000C00000076342E302E33303331390000000005006C000000DC030000237E0000480400003803000023537472696E67730000000080070000800000002355530000080000100000002347554944000000100800002401000023426C6F6200000000000000020000015717A201090A000000FA013300160000010000000E000000040000000700000010000000190000000100000011000000070000000300000001000000040000000800000001000000030000000200000002000000000092010100000000000600FF00580206002E0158020600EB0024020F00780200000A00D802BA020E00DA0137020A009F00BA020A00AA02BA0206009500C1010A001F01BA020A006900BA020A00C600BA0206006701C1010600E602C101000000002900000000000100010001001000DB0100001500010001000A0110007A0100002500010011000A0110006E0100002500040011000600D0017F000600F1017F000600FE028300060012027F00060002027F000600ED0283000600FE02830050200000000086181E02860001008920000000008608A401900006009B20000000008108AF0194000600AE200000000086087E0090000700C020000000008108870094000700D320000000008608870299000800E52000000000810894029D000800F82000000000E6090C03A20009000A210000000081081C03A60009001D2100000000E6014900AB000A003A210000000081005400B1000B00582100000000E601E901B8000D000C2200000000E601C801BF000F00C42200000000E6015201C800120035230000000081008601D0001500592300000000E6015F00D900180000000100B300000002002C0300000300900000000400BA0100000500A102000001004C01000001004C01000001004C01000001004C0100000100B20200000100B202000002004C0100000100F40100000200050300000100D50100000200F401000003000503000001000A0200000200F702000003000503000001001802000002000A02000003004C01000001001802000002000A020200190009001E02010011001E02060019001E020A0051001E02060029001E0210002900720016005900F7011B002900D700200061005A012500610064012A0061000100300061000B003500610015003B0061001F0040006900D10246002900820152006900D1026F00210023001E012E000B00ED002E001300F6002E001B001501410023001E01810023001E01A10023001E014C005E006500020001000000B301E10000008B00E10000009802E50000002003E900020002000300010003000300020004000500010005000500020006000700010007000700020008000900010009000900048000000000000000000000000000000000D802000004000000000000000000000076003200000000000100020001000000000000000000BA0200000000010000000000000000000000000037020000000003000200040002002100590021006A0000000047657455496E7433320053657455496E7433320047657455496E7436340053657455496E743634003C4D6F64756C653E0053797374656D2E507269766174652E436F72654C69620047657442616C616E63650053657442616C616E636500416C6C6F77616E636500494D657373616765006765745F4D657373616765006765745F4E616D65007365745F4E616D65006E616D650056616C7565547970650049536D617274436F6E7472616374537461746500736D617274436F6E74726163745374617465004950657273697374656E745374617465006765745F50657273697374656E7453746174650044656275676761626C6541747472696275746500436F6D70696C6174696F6E52656C61786174696F6E7341747472696275746500496E6465784174747269627574650052756E74696D65436F6D7061746962696C6974794174747269627574650076616C756500417070726F766500476574537472696E6700536574537472696E6700417070726F76616C4C6F67005472616E736665724C6F6700536574417070726F76616C00536D617274436F6E74726163742E646C6C006765745F53796D626F6C007365745F53796D626F6C0073796D626F6C0053797374656D005472616E7366657246726F6D0066726F6D00495374616E64617264546F6B656E005472616E73666572546F00746F006765745F53656E646572005370656E646572007370656E646572004F776E6572006F776E6572002E63746F720053797374656D2E446961676E6F737469637300537472617469732E536D617274436F6E7472616374732E5374616E64617264730053797374656D2E52756E74696D652E436F6D70696C6572536572766963657300446562756767696E674D6F646573006765745F446563696D616C73007365745F446563696D616C7300646563696D616C730041646472657373006164647265737300537472617469732E536D617274436F6E74726163747300466F726D617400536D617274436F6E7472616374004F626A656374004F6C64416D6F756E740063757272656E74416D6F756E7400616D6F756E74006765745F546F74616C537570706C79007365745F546F74616C537570706C7900746F74616C537570706C7900000D530079006D0062006F006C0000094E0061006D006500001144006500630069006D0061006C007300001754006F00740061006C0053007500700070006C0079000017420061006C0061006E00630065003A007B0030007D00002341006C006C006F00770061006E00630065003A007B0030007D003A007B0031007D000000DA594A134F8C284CA28DA6D4022688B40004200101080320000105200101111105200101121D042000122D042000112104200012310420010E0E052002010E0E042001090E052002010E090420010B0E052002010E0B0500020E0E1C0507020B110C06300101011E00040A01110C0607030B0B110C0407011110040A0111100600030E0E1C1C087CEC85D7BEA7798E0306112102060B09200501121D0B0E0E090320000E042001010E0320000904200101090320000B042001010B0520010B11210620020111210B0620020211210B08200302112111210B0720030211210B0B08200301112111210B0720020B112111210328000E032800090328000B0801000800000000001E01000100540216577261704E6F6E457863657074696F6E5468726F77730108010002000000000004010000000000000000000000000000000010000000000000000000000000000000F42C000000000000000000000E2D0000002000000000000000000000000000000000000000000000002D0000000000000000000000005F436F72446C6C4D61696E006D73636F7265652E646C6C0000000000FF250020001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000C000000203D00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

YakupIpek commented on Mainnet/StandardToken/StandardToken.Tests/StandardTokenTests.cs:
```diff
@@ -580,18 +597,19 @@ public void TransferFrom_Self()
 
         [Fact]
-        public void Constructor_Sets_Name_And_Symbol()
+        public void Constructor_Sets_Name_Symbol_And_Decimals()
```
Can you update this test for having all parameters ? Currently only TotalSupply is missing. Also give it a generic name `Constructor_Sets_Parameters` after that you can remove `Constructor_Sets_TotalSupply`

mrtpain commented on Mainnet/StandardToken/StandardToken/StandardToken.cs:
```diff
@@ -52,6 +54,16 @@ private void SetBalance(Address address, ulong value)
+        PersistentState.SetUInt32("Decimals", value);
+    }
+
     /// <inheritdoc />
```
I'll leave this decision up to you and the team. Couple points to discuss and reasoning behind having methods instead.

1. This was initially added with Get/Set methods so the Get could be called by another contract (before we had uint256 and I needed it for math operations to safely multiply large numbers)
2. Looking back at `IStandardToken` we just added `GetDecimals` to the interface, will need to update that if we remove the method and exclusively use the property.
3. ERC20 contracts have `Name`, `Symbol`, and `Decimals` as properties but they also all have public `Get()` methods so other contracts can access these values from a `Call()`

You and the team can decide on the right approach, again, my opinion is to side with ERC standards right now. Which would include:
- Adding the Decimals property like you suggested
- Keeping the GetDecimals() method
- Removing the SetDecimals method and just using `Decimals = decimals` in the constructor
- Adding a GetName() method
- Adding a GetSymbol method
- Adding a GetTotalSupply method

In reference to Open Zeppelin ERC20 contract: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L55-L98

mrtpain commented:
@YakupIpek Updated, adjusted tests and added README

For future reference, the decision was to use a `Decimals` property and contracts can make a call to get the value by using:
```C#
Call(ContractAddress, 0, $"get_{nameof(Property)}", null);
```

I lumped Testnet in this PR as well now so Testnet and Mainnet are identical.

quantumagi commented:
This should probably not be merged until a sufficient number of nodes have upgraded to PR #365.

mrtpain commented:
Closing now that this is a dup PR - https://github.com/stratisproject/CirrusSmartContracts/pull/32
